src/python/pants/backend/python/rules/resolve.py

```diff
@@ -0,0 +1,356 @@
+    globs_by_resolve_result: DefaultDict[ResolveResult, List[str]] = defaultdict(list)
+
+    for resolved_dependency in resolved_dependencies:
+        # Include everything in the cache by default.
+        globs_by_resolve_result[resolved_dependency].append(
+            f"{resolved_dependency.cache.path}/**/*"
+        )
+        # But flag any files under http/ as potential conflicts since we know the Pip HTTP cache
+        # has non-deterministic headers.
+        potential_conflict = re.compile(rf"{resolved_dependency.cache.path}/http/")
+        for path in resolved_dependency.cache.snapshot.files:
+            if potential_conflict.match(path):
+                resolve_results_by_path[path].append(resolved_dependency)
+
+    potential_conflicts: Dict[str, List[ResolveResult]] = {
+        path: results[1:] for path, results in resolve_results_by_path.items() if len(results) > 1
+    }
+    for path, resolve_results in potential_conflicts.items():
+        for resolve_result in resolve_results:
+            globs_by_resolve_result[resolve_result].append(f"!{path}")
+
+    mergeable_caches = await MultiGet(
+        Get[Snapshot](
+            SnapshotSubset(
+                digest=resolved_dependency.cache.snapshot.digest,
+                globs=PathGlobs(globs=globs_by_resolve_result[resolved_dependency]),
+            )
+        )
+        for resolved_dependency in resolved_dependencies
+    )
+    merged_cache_snapshot = await Get[Snapshot](
+        MergeDigests(digests=tuple(mergeable_cache.digest for mergeable_cache in mergeable_caches))
+    )
+
+    return await Get[ResolveResult](
+        ResolveRequest(
+            requirer=target.address.spec,
+            requirements=PexRequirements(
+                tuple(
+                    str(requirement.requirement)
+                    for target in transitive_dependencies.closure
+                    if isinstance(target, PythonRequirementLibrary)
+                    for requirement in target[PythonRequirementsField].value
+                )
+            ),
+            cache=PexCache(snapshot=merged_cache_snapshot),
+            local=options.local,
+        )
+    )
+
+
+class ResolveGoal(Goal):
+    """Resolves third party dependencies for a target."""
+
+    subsystem_cls = ResolveOptions
+
+
+logger = logging.getLogger(__name__)
+
+
+@goal_rule
+async def resolve(
+    targets: Targets,
+    distdir: DistDir,
+    workspace: Workspace,
+    options: ResolveOptions,
+    console: Console,
+) -> ResolveGoal:
+
+    if len(targets) != 1:
+        raise ValueError(
+            f"The `{options.name}` goal can currently only work with 1 target but given "
+            f"{len(targets)} targets."
+        )
+    target = targets[0]
+    if not isinstance(target, PythonLibrary):
+        raise ValueError(
+            f"The `{options.name}` goal currently only works with "
+            f"`{PythonLibrary.alias}` targets but given a `{target.alias}` target."
+        )
```

Comment:
If you fix here, It looks like attributes you care about are a) Python code, and b) `Dependencies`. If so, use this:

```python
class ResolveFieldSet(FieldSet):
   required_fields = (PythonSources,)

   dependencies: Dependencies
```

This means that we never will actually consume `PythonSources`. We only use it to filter out irrelevant targets. We also will use `tgt.get(Depenencies)` instead of `tgt[Dependencies]`, as the default `Dependencies(raw_value=None)` works just fine if that target type doesn't have the field.

--

Then, your rule will request `ResolveFieldSet`. It consumes it with `field_set.address` and `field_set.dependencies.`